File: spec/screens/online_match_model_spec.lua
-- Tier-2 coverage for the pure online match flow model (AGENTS.md §9).
--
-- `spec/screens/online_match_flow_spec.lua` drives this model end to end through
-- two mounted screens, which proves it works in the real flow but exercises it
-- only through whatever a real match happens to produce. This spec calls
-- `online_match_model.command` directly, one event kind at a time, against
-- hand-built states — which is the point of splitting a screen into a pure model
-- and a thin shell in the first place.
--
-- Two deliberate shortcuts, both of which keep the subject under test honest:
--
--  * The coordinator states come from `game.online.coordinator_driver`, real
--    instances driven to `running` over real canonical wires. That driver
--    proposes the *protocol fixture* manifest, while the request here is built
--    from the content-derived one. The model never reconciles the two — it reads
--    only the request's role, peer id, and first input tick — so pairing them is
--    safe, and it keeps this spec from re-testing the lobby.
--  * `RollbackEventStep`s are hand-built with only the fields the model reads
--    (`tick` and `lifecycle_events`). That *is* the model's input surface; a full
--    simulated step would add nothing but coupling.

local t = require("spec.support.runner")
local online_fixture = require("spec.fixtures.online_match_session")
local coordinator_driver = require("game.online.coordinator_driver")
local coordinator_fixture = require("game.online.coordinator_fixture")
local match_session = require("game.online.match_session")
local online_match_model = require("game.screens.online_match_model")
local protocol = require("game.online.protocol")

local MODE = "1v1"
-- The coordinator driver's own peer ids, so the freeze names the same humans the
-- coordinator instances do.
local SESSION_PEERS = { coordinator_fixture.HOST_PEER_ID, coordinator_fixture.guest_peer_id(1) }

---@class OnlineModelHarness
---@field host OnlineMatchModel
---@field guest OnlineMatchModel

---@return OnlineModelHarness
local function harness()
    local manifest = online_fixture.manifest(MODE)
    local freeze = online_fixture.freeze(manifest, SESSION_PEERS)
    local session = coordinator_driver.new({ guest_count = 1, mode = MODE }):reach_start()
    local nodes = session.nodes
    assert(nodes[1].state.phase == "running", "the host never reached running")
    assert(nodes[2].state.phase == "running", "the guest never reached running")

    ---@param index integer
    ---@param role SessionRole
    ---@return OnlineMatchModel
    local function model_for(index, role)
        local request = assert(match_session.request({
            role = role,
            peer_id = SESSION_PEERS[index],
            manifest = manifest,
            freeze = freeze,
        }))
        return online_match_model.new(request, nodes[index].state)
    end

    return { host = model_for(1, "host"), guest = model_for(2, "guest") }
end

---@param tick integer
---@param lifecycle RollbackWrappedLifecycleEvent[]?
---@return RollbackEventStep
local function step(tick, lifecycle)
    return {
        tick = tick,
        start_boundary = tick,
        end_boundary = tick + 1,
        state = { score = { home = 0, away = 0 }, time_left = 1, finished = false },
        match_events = {},
        combat_events = {},
        lifecycle_events = lifecycle or {},
    }
end

---@param kind RollbackLifecycleKind
---@param home integer
---@param away integer
---@param team InputTeam?
---@return RollbackWrappedLifecycleEvent
local function lifecycle(kind, home, away, team)
    return {
        id = "lifecycle/" .. kind .. "/1",
        tick = 0,
        domain = "lifecycle/" .. kind,
        ordinal = 1,
        payload = { kind = kind, team = team, score = { home = home, away = away } },
    }
end

---@param effects OnlineMatchEffect[]
---@return SessionMatchPhase[]
local function published_phases(effects)
    ---@type SessionMatchPhase[]
    local phases = {}
    for _, effect in ipairs(effects) do
        if effect.kind == "send" then
            local message = protocol.decode(effect.wire)
            if message and message.kind == "match_phase" then
                local body = message.body
                ---@cast body SessionMatchPhaseBody
                phases[#phases + 1] = body.phase
            end
        end
    end
    return phases
end

---@param effects OnlineMatchEffect[]
---@param kind string
---@return boolean
local function sent(effects, kind)
    for _, effect in ipairs(effects) do
        if effect.kind == "send" then
            local message = protocol.decode(effect.wire)
            if message and message.kind == kind then
                return true
            end
        end
    end
    return false
end

t.describe("online match model purity", function()
    t.it("returns a fresh model and never mutates the one it was given", function()
        local model = harness().host
        local before = model.coordinator
        local next_model, effects = online_match_model.command(model, { kind = "focus_lost" })
        t.is_true(next_model ~= model, "command must copy")
        t.eq(model.notice, nil, "the input model must be untouched")
        t.eq(model.abort_prompt, false)
        t.is_true(next_model.notice ~= nil)
        t.eq(#effects, 0, "a focus notice is local and sends nothing")
        t.is_true(next_model.coordinator == before, "a notice never steps the session")
    end)

    t.it("ignores an unknown event kind rather than failing", function()
        local model = harness().host
        local next_model, effects = online_match_model.command(model, { kind = "not_a_real_event" })
        t.eq(#effects, 0)
        t.eq(next_model.phase, "playing")
        t.eq(next_model.error, nil)
    end)
end)

t.describe("online match model local interruptions", function()
    -- The whole point: none of these may stop the driver, so none of them may
    -- produce a coordinator effect either.
    t.it("answers focus loss and controller loss with a notice and nothing else", function()
        for _, case in ipairs({
            { kind = "focus_lost", text = online_match_model.FOCUS_NOTICE },
            { kind = "controller_lost", text = online_match_model.CONTROLLER_NOTICE },
        }) do
            local model = harness().host
            local next_model, effects = online_match_model.command(model, { kind = case.kind })
            t.eq(#effects, 0, case.kind .. " must not reach the session")
            t.eq(assert(next_model.notice).text, case.text)
            t.eq(next_model.phase, "playing")
            t.eq(next_model.abort_prompt, false, case.kind .. " must not arm the abort prompt")
        end
    end)

    t.it("explains a pause request before it will abort", function()
        local model = harness().host
        local first, effects = online_match_model.command(model, { kind = "pause_request" })
        t.eq(#effects, 0, "the first pause request only explains itself")
        t.is_true(first.abort_prompt)
        t.eq(assert(first.notice).text, online_match_model.PAUSE_NOTICE)
        t.eq(first.phase, "playing")
        t.eq(first.terminal, nil)
    end)

    t.it("aborts on a second pause request and ends the session", function()
        local model = harness().host
        local armed = (online_match_model.command(model, { kind = "pause_request" }))
        local aborted, effects = online_match_model.command(armed, { kind = "pause_request" })
        t.is_true(sent(effects, "abort"), "an abort must be announced to the peers")
        t.eq(aborted.phase, "ended")
        t.eq(assert(aborted.terminal).reason, "local_abort")
        t.eq(aborted.result, nil, "an aborted session has no result to commit")
        t.eq(online_match_model.exit_route(aborted), "terminal")
        t.is_true(online_match_model.ended(aborted))
    end)

    t.it("disarms the prompt on any other input", function()
        local model = harness().host
        local armed = (online_match_model.command(model, { kind = "pause_request" }))
        local resumed, effects = online_match_model.command(armed, { kind = "resume_request" })
        t.eq(#effects, 0)
        t.eq(resumed.abort_prompt, false)
        t.eq(resumed.phase, "playing")
        t.eq(resumed.terminal, nil)
    end)

    t.it("decays a notice on the session clock", function()
        local model = harness().host
        local noticed = (online_match_model.command(model, { kind = "focus_lost" }))
        t.eq(assert(noticed.notice).remaining, online_match_model.NOTICE_SECONDS)
        local ticked = (online_match_model.command(noticed, { kind = "tick", dt = 1 }))
        t.eq(assert(ticked.notice).remaining, online_match_model.NOTICE_SECONDS - 1)
        local expired = (
            online_match_model.command(ticked, {
                kind = "tick",
                dt = online_match_model.NOTICE_SECONDS,
            })
        )
        t.eq(expired.notice, nil)
    end)
end)

t.describe("online match model confirmed publication", function()
    t.it("opens the session with kickoff then playing on the first confirmed step", function()
        local model = harness().host
        local next_model, effects = online_match_model.command(model, {
            kind = "confirmed",
            steps = { step(0) },
        })
        t.eq(table.concat(published_phases(effects), ","), "kickoff,playing")
        t.eq(assert(next_model.coordinator.match).phase, "playing")
        t.eq(next_model.phase, "playing")
    end)

    t.it("publishes a goal stoppage only after the confirmed score moved", function()
        local model = harness().host
        model = (online_match_model.command(model, { kind = "confirmed", steps = { step(0) } }))
        local scored, effects = online_match_model.command(model, {
            kind = "confirmed",
            steps = { step(1, { lifecycle("goal", 1, 0, "home"), lifecycle("kickoff", 1, 0) }) },
        })
        t.eq(table.concat(published_phases(effects), ","), "goal_stoppage,kickoff,playing")
        t.eq(scored.score.home, 1)
        t.eq(scored.score.away, 0)
        local match = assert(scored.coordinator.match)
        t.eq(match.home_score, 1, "the session carries the confirmed score, not a predicted one")
        t.eq(match.phase, "playing")
    end)

    t.it("walks the legal phase chain rather than publishing an illegal transition", function()
        local model = harness().host
        model = (online_match_model.command(model, { kind = "confirmed", steps = { step(0) } }))
        -- A goal leaves the session in `goal_stoppage`; full time cannot follow it
        -- directly, so the model has to route back through kickoff and playing.
        model = (
            online_match_model.command(model, {
                kind = "confirmed",
                steps = { step(1, { lifecycle("goal", 1, 0, "home") }) },
            })
        )
        t.eq(model.published, "playing")
        t.eq(model.error, nil, "no phase may be refused as out of order")
    end)

    t.it("never lets a guest publish a match phase", function()
        local model = harness().guest
        local next_model, effects = online_match_model.command(model, {
            kind = "confirmed",
            steps = { step(0, { lifecycle("goal", 1, 0, "home") }) },
        })
        t.eq(#published_phases(effects), 0, "only the host publishes match phases")
        t.eq(next_model.published, nil)
        t.eq(next_model.score.home, 0, "a guest does not restate the score either")
    end)

    t.it("reports a boundary checkpoint as a hash report", function()
        local model = harness().host
        local _, effects = online_match_model.command(model, {
            kind = "checkpoints",
            checkpoints = { { tick = 30, hash = "0123456789abcdef", live = {} } },
        })
        t.is_true(sent(effects, "hash_report"))
    end)

    t.it("surfaces a peer's boundary hash to the driver as well as the session", function()
        local model = harness().host
        local wire = assert(
            protocol.encode(
                assert(
                    protocol.new(
                        "hash_report",
                        model.coordinator.session_id,
                        coordinator_fixture.guest_peer_id(1),
                        9,
                        { tick = 30, boundary_hash = "0123456789abcdef" }
                    )
                )
            )
        )
        local _, effects = online_match_model.command(model, {
            kind = "control",
            link_id = coordinator_fixture.link_id(coordinator_fixture.guest_peer_id(1)),
            wire = wire,
        })
        local observed = nil
        for _, effect in ipairs(effects) do
            if effect.kind == "observe_hash" then
                observed = effect
            end
        end
        t.eq(assert(observed).tick, 30)
        t.eq(assert(observed).hash, "0123456789abcdef")
    end)
end)

t.describe("online match model full time and result", function()
    ---@param model OnlineMatchModel
    ---@return OnlineMatchModel, OnlineMatchEffect[]
    local function reach_full_time(model)
        model = (online_match_model.command(model, { kind = "confirmed", steps = { step(0) } }))
        return online_match_model.command(model, {
            kind = "full_time",
            final_tick = 90,
            home_score = 2,
            away_score = 1,
            final_hash = "0123456789abcdef",
        })
    end

    t.it("walks to full time and reports the simulation's own result", function()
        local finished, effects = reach_full_time(harness().host)
        t.eq(table.concat(published_phases(effects), ","), "full_time,result")
        t.is_true(sent(effects, "result_ack"), "the host offers the result for acknowledgement")
        t.eq(finished.phase, "full_time")
        t.eq(finished.score.home, 2)
        t.eq(finished.score.away, 1)
    end)

    t.it("does not commit a result while any peer has not acknowledged it", function()
        local finished = (reach_full_time(harness().host))
        t.eq(finished.result, nil, "the result waits for the session, not the simulation")
        t.eq(finished.phase, "full_time")
        t.is_true(not online_match_model.ended(finished))
    end)

    t.it("makes a guest report its own result rather than accept one on faith", function()
        local finished, effects = reach_full_time(harness().guest)
        t.eq(#published_phases(effects), 0, "a guest publishes no match phase")
        t.eq(finished.phase, "full_time")
        -- The guest's `finish` is local: it becomes `local_result`, which the
        -- coordinator later compares against the host's acknowledgement.
        local local_result = assert(finished.coordinator.local_result)
        t.eq(local_result.home_score, 2)
        t.eq(local_result.away_score, 1)
        t.eq(local_result.final_hash, "0123456789abcdef")
        t.eq(finished.result, nil)
    end)

    t.it("ignores a second full time", function()
        local first = (reach_full_time(harness().host))
        local second, effects = online_match_model.command(first, {
            kind = "full_time",
            final_tick = 91,
            home_score = 9,
            away_score = 9,
            final_hash = "ffffffffffffffff",
        })
        t.eq(#effects, 0, "full time happens once")
        t.eq(second.score.home, 2)
        t.eq(second.score.away, 1)
    end)
end)

t.describe("online match model terminal reporting", function()
    t.it("reports a netcode failure class into the session", function()
        for _, case in ipairs({
            { failure = "late_input", reason = "late_input" },
            { failure = "desync", reason = "hash_mismatch" },
            { failure = "input_channel", reason = "input_channel_failure" },
        }) do
            local model = harness().host
            local next_model = (
                online_match_model.command(model, {
                    kind = "failure",
                    status = "late_input",
                    failure = case.failure,
                    detail = "a driver terminal",
                })
            )
            t.eq(assert(next_model.terminal).reason, case.reason, case.failure)
            t.eq(next_model.phase, "ended")
            t.eq(next_model.result, nil, "a failed session commits no score")
            t.eq(online_match_model.exit_route(next_model), "terminal")
        end
    end)

    t.it("reports a lost star as the guest's own lost host link", function()
        local model = harness().guest
        local next_model, effects = online_match_model.command(model, {
            kind = "failure",
            status = "transport_lost",
            failure = nil,
            detail = "the star transport is no longer connected",
        })
        t.eq(assert(next_model.terminal).reason, "transport_lost")
        t.eq(next_model.phase, "ended")
        t.eq(#published_phases(effects), 0)
    end)

    t.it("ends a host session explicitly when it cannot name the lost peer", function()
        local model = harness().host
        local next_model, effects = online_match_model.command(model, {
            kind = "failure",
            status = "transport_lost",
            failure = nil,
            detail = "the star transport is no longer connected",
        })
        -- The driver folds every link event into one status, so a host has no
        -- peer id to disconnect. It ends the session rather than guessing.
        t.eq(assert(next_model.terminal).reason, "local_abort")
        t.is_true(sent(effects, "abort"))
        t.eq(next_model.phase, "ended")
    end)

    t.it("keeps the first terminal reason when late traffic follows it", function()
        local model = harness().host
        local ended = (
            online_match_model.command(model, {
                kind = "failure",
                status = "late_input",
                failure = "late_input",
            })
        )
        local after, effects = online_match_model.command(ended, {
            kind = "confirmed",
            steps = { step(0) },
        })
        t.eq(#effects, 0, "a terminated session makes no further progress")
        t.eq(assert(after.terminal).reason, "late_input")
    end)
end)

File: ts/packages/screens/src/online_match_model.spec.ts
// This suite exercises `online_match_model.ts`'s pure flow policy. Its
// `CoordinatorState` instances never come from a real, running session: the
// coordinator itself is Rust-owned (`crates/gc-netcode`; ARCHITECTURE.md
// §1.1) with no wasm bridge yet, so there is no way to drive one to
// `running` over real canonical wires here.
//
// That's not a loss for this suite's purpose. Every assertion below is
// about `online_match_model.ts`'s *own* control flow: which coordinator
// events it issues, in what order, and how its local fields (`phase`,
// `score`, `tick`, `notice`, `abort_prompt`, `terminal`, `result`) evolve.
// The coordinator itself is only ever asked to relay a message or flip to
// `terminal`, never to arbitrate admission, slots, or readiness -- so a
// small hand-scripted `CoordinatorPort` fake (in the same spirit as
// `match_presentation.spec.ts`'s `fakeRollbackEvents`) is enough to run
// every test below for real, without reimplementing any of the
// coordinator's own protocol logic.

import { describe, expect, it } from "vitest";
import {
  CONTROLLER_NOTICE,
  FOCUS_NOTICE,
  NOTICE_SECONDS,
  PAUSE_NOTICE,
  command,
  ended,
  exitRoute,
  newOnlineMatchModel,
  type CoordinatorEvent,
  type CoordinatorOutcome,
  type CoordinatorPort,
  type CoordinatorStateCore,
  type CoordinatorTerminal,
  type OnlineMatchEffect,
  type OnlineMatchModel,
  type OnlineMatchModelPorts,
  type OnlineMatchRequestCore,
  type ProtocolMessage,
  type ProtocolPort,
  type RollbackEventStepCore,
  type RollbackWrappedEvent,
  type LifecyclePayload,
} from "./online_match_model.ts";

// ---------------------------------------------------------------------------
// Fake coordinator/protocol: message relay + terminal flip only, no protocol
// acceptance/admission/readiness logic. See this file's header.
// ---------------------------------------------------------------------------

interface FakeCoordinatorState extends CoordinatorStateCore {
  readonly session_id: string;
  readonly local_result?: {
    readonly home_score: number;
    readonly away_score: number;
    readonly final_hash: string;
  };
}

function fakeCoordinator(role: "host" | "guest", hostLinkId?: string): FakeCoordinatorState {
  return {
    phase: "running",
    role,
    session_id: "session/fixture",
    ...(hostLinkId !== undefined ? { host_link_id: hostLinkId } : {}),
  };
}

function terminalReasonFor(event: CoordinatorEvent): string {
  if (event["kind"] === "abort") {
    return "local_abort";
  }
  if (event["kind"] === "link_lost") {
    // `code` is a `TransportErrorCode` (a string union) on this event; read it
    // as one rather than stringifying an `unknown`, which would silently
    // produce "[object Object]" if the shape ever changed.
    const code = event["code"];
    return typeof code === "string" ? code : "transport_lost";
  }
  if (event["kind"] === "netcode_failure") {
    const failure = event["failure"];
    if (failure === "late_input") return "late_input";
    if (failure === "desync") return "hash_mismatch";
    if (failure === "input_channel") return "input_channel_failure";
    return "unknown";
  }
  throw new Error(`fake coordinator cannot terminate on ${String(event["kind"])}`);
}

function fakeCoordinatorPort(): CoordinatorPort<FakeCoordinatorState> {
  return {
    step(state, event): readonly [FakeCoordinatorState, CoordinatorOutcome] {
      switch (event["kind"]) {
        case "abort":
        case "link_lost":
        case "netcode_failure": {
          const terminal: CoordinatorTerminal = {
            reason: terminalReasonFor(event),
            ...(typeof event["detail"] === "string" ? { detail: event["detail"] } : {}),
          };
          const outcome: CoordinatorOutcome = {
            accepted: true,
            actions: [
              { kind: "send", message: event, targets: ["peer"] },
              { kind: "terminate", terminal },
            ],
          };
          return [{ ...state, phase: "terminal" }, outcome];
        }
        case "match_phase": {
          const outcome: CoordinatorOutcome = {
            accepted: true,
            actions: [
              {
                kind: "send",
                message: { kind: "match_phase", body: { phase: event["phase"] } },
                targets: ["peer"],
              },
            ],
          };
          return [state, outcome];
        }
        case "hash_report": {
          const outcome: CoordinatorOutcome = {
            accepted: true,
            actions: [
              {
                kind: "send",
                message: {
                  kind: "hash_report",
                  body: { tick: event["tick"], boundary_hash: event["boundary_hash"] },
                },
                targets: ["peer"],
              },
            ],
          };
          return [state, outcome];
        }
        case "finish": {
          if (state.role === "host") {
            const outcome: CoordinatorOutcome = {
              accepted: true,
              actions: [
                {
                  kind: "send",
                  message: {
                    kind: "result_ack",
                    body: { home_score: event["home_score"], away_score: event["away_score"] },
                  },
                  targets: ["peer"],
                },
              ],
            };
            return [state, outcome];
          }
          const localResult = {
            home_score: event["home_score"] as number,
            away_score: event["away_score"] as number,
            final_hash: event["final_hash"] as string,
          };
          return [
            { ...state, local_result: localResult },
            { accepted: true, actions: [] },
          ];
        }
        default:
          return [state, { accepted: true, actions: [] }];
      }
    },
  };
}

function fakeProtocolPort(): ProtocolPort {
  return {
    encode(message): string {
      return JSON.stringify(message);
    },
    decode(wire): ProtocolMessage | undefined {
      try {
        return JSON.parse(wire) as ProtocolMessage;
      } catch {
        return undefined;
      }
    },
  };
}

function ports(): OnlineMatchModelPorts<FakeCoordinatorState> {
  return {
    coordinator: fakeCoordinatorPort(),
    protocol: fakeProtocolPort(),
    localResult: () => undefined,
    isCompleted: (terminal) => terminal.reason === "completed",
  };
}

interface FixtureRequest extends OnlineMatchRequestCore {
  readonly peer_id: string;
}

function harness() {
  const request = (role: "host" | "guest"): FixtureRequest => ({
    role,
    peer_id: role,
    first_input_tick: 0,
  });
  return {
    host: newOnlineMatchModel<FakeCoordinatorState, FixtureRequest>(
      request("host"),
      fakeCoordinator("host"),
    ),
    guest: newOnlineMatchModel<FakeCoordinatorState, FixtureRequest>(
      request("guest"),
      fakeCoordinator("guest", "host_link"),
    ),
  };
}

function step(
  tick: number,
  lifecycle?: readonly RollbackWrappedEvent<LifecyclePayload>[],
): RollbackEventStepCore {
  return { tick, lifecycle_events: lifecycle ?? [] };
}

function lifecycle(
  kind: "goal" | "kickoff" | "full_time",
  home: number,
  away: number,
  team?: "home" | "away",
): RollbackWrappedEvent<LifecyclePayload> {
  const payload: LifecyclePayload =
    kind === "goal"
      ? { kind: "goal", team: team ?? "home", score: { home, away } }
      : kind === "kickoff"
        ? { kind: "kickoff" }
        : { kind: "full_time" };
  return { id: `lifecycle/${kind}/1`, tick: 0, domain: `lifecycle/${kind}`, ordinal: 1, payload };
}

function publishedPhases(effects: readonly OnlineMatchEffect[]): string[] {
  const phases: string[] = [];
  for (const effect of effects) {
    if (effect.kind === "send") {
      const message = fakeProtocolPort().decode(effect.wire);
      if (message?.kind === "match_phase") {
        phases.push(String((message.body as { readonly phase: string }).phase));
      }
    }
  }
  return phases;
}

function sent(effects: readonly OnlineMatchEffect[], kind: string): boolean {
  return effects.some(
    (effect) => effect.kind === "send" && fakeProtocolPort().decode(effect.wire)?.kind === kind,
  );
}

describe("online match model purity", () => {
  it("returns a fresh model and never mutates the one it was given", () => {
    const model = harness().host;
    const before = model.coordinator;
    const [next, effects] = command(model, ports(), { kind: "focus_lost" });
    expect(next).not.toBe(model);
    expect(model.notice).toBeUndefined();
    expect(model.abort_prompt).toBe(false);
    expect(next.notice).toBeDefined();
    expect(effects.length).toBe(0);
    expect(next.coordinator).toBe(before);
  });

  it("ignores an unknown event kind rather than failing", () => {
    const model = harness().host;
    const [next, effects] = command(model, ports(), { kind: "not_a_real_event" });
    expect(effects.length).toBe(0);
    expect(next.phase).toBe("playing");
    expect(next.error).toBeUndefined();
  });
});

describe("online match model local interruptions", () => {
  it("answers focus loss and controller loss with a notice and nothing else", () => {
    for (const testCase of [
      { kind: "focus_lost" as const, text: FOCUS_NOTICE },
      { kind: "controller_lost" as const, text: CONTROLLER_NOTICE },
    ]) {
      const model = harness().host;
      const [next, effects] = command(model, ports(), { kind: testCase.kind });
      expect(effects.length).toBe(0);
      expect(next.notice?.text).toBe(testCase.text);
      expect(next.phase).toBe("playing");
      expect(next.abort_prompt).toBe(false);
    }
  });

  it("explains a pause request before it will abort", () => {
    const model = harness().host;
    const [first, effects] = command(model, ports(), { kind: "pause_request" });
    expect(effects.length).toBe(0);
    expect(first.abort_prompt).toBe(true);
    expect(first.notice?.text).toBe(PAUSE_NOTICE);
    expect(first.phase).toBe("playing");
    expect(first.terminal).toBeUndefined();
  });

  it("aborts on a second pause request and ends the session", () => {
    const model = harness().host;
    const [armed] = command(model, ports(), { kind: "pause_request" });
    const [aborted, effects] = command(armed, ports(), { kind: "pause_request" });
    expect(sent(effects, "abort")).toBe(true);
    expect(aborted.phase).toBe("ended");
    expect(aborted.terminal?.reason).toBe("local_abort");
    expect(aborted.result).toBeUndefined();
    expect(exitRoute(aborted)).toBe("terminal");
    expect(ended(aborted)).toBe(true);
  });

  it("disarms the prompt on any other input", () => {
    const model = harness().host;
    const [armed] = command(model, ports(), { kind: "pause_request" });
    const [resumed, effects] = command(armed, ports(), { kind: "resume_request" });
    expect(effects.length).toBe(0);
    expect(resumed.abort_prompt).toBe(false);
    expect(resumed.phase).toBe("playing");
    expect(resumed.terminal).toBeUndefined();
  });

  it("decays a notice on the session clock", () => {
    const model = harness().host;
    const [noticed] = command(model, ports(), { kind: "focus_lost" });
    expect(noticed.notice?.remaining).toBe(NOTICE_SECONDS);
    const [ticked] = command(noticed, ports(), { kind: "tick", dt: 1 });
    expect(ticked.notice?.remaining).toBe(NOTICE_SECONDS - 1);
    const [expired] = command(ticked, ports(), { kind: "tick", dt: NOTICE_SECONDS });
    expect(expired.notice).toBeUndefined();
  });
});

describe("online match model confirmed publication", () => {
  it("opens the session with kickoff then playing on the first confirmed step", () => {
    const model = harness().host;
    const [next, effects] = command(model, ports(), { kind: "confirmed", steps: [step(0)] });
    expect(publishedPhases(effects).join(",")).toBe("kickoff,playing");
    expect(next.phase).toBe("playing");
  });

  it("publishes a goal stoppage only after the confirmed score moved", () => {
    let model = harness().host;
    [model] = command(model, ports(), { kind: "confirmed", steps: [step(0)] });
    const [scored, effects] = command(model, ports(), {
      kind: "confirmed",
      steps: [step(1, [lifecycle("goal", 1, 0, "home"), lifecycle("kickoff", 1, 0)])],
    });
    expect(publishedPhases(effects).join(",")).toBe("goal_stoppage,kickoff,playing");
    expect(scored.score.home).toBe(1);
    expect(scored.score.away).toBe(0);
  });

  it("walks the legal phase chain rather than publishing an illegal transition", () => {
    let model = harness().host;
    [model] = command(model, ports(), { kind: "confirmed", steps: [step(0)] });
    [model] = command(model, ports(), {
      kind: "confirmed",
      steps: [step(1, [lifecycle("goal", 1, 0, "home")])],
    });
    expect(model.published).toBe("playing");
    expect(model.error).toBeUndefined();
  });

  it("never lets a guest publish a match phase", () => {
    const model = harness().guest;
    const [next, effects] = command(model, ports(), {
      kind: "confirmed",
      steps: [step(0, [lifecycle("goal", 1, 0, "home")])],
    });
    expect(publishedPhases(effects).length).toBe(0);
    expect(next.published).toBeUndefined();
    expect(next.score.home).toBe(0);
  });

  it("reports a boundary checkpoint as a hash report", () => {
    const model = harness().host;
    const [, effects] = command(model, ports(), {
      kind: "checkpoints",
      checkpoints: [{ tick: 30, hash: "0123456789abcdef" }],
    });
    expect(sent(effects, "hash_report")).toBe(true);
  });

  it("surfaces a peer's boundary hash to the driver as well as the session", () => {
    const model = harness().host;
    const wire = fakeProtocolPort().encode({
      kind: "hash_report",
      body: { tick: 30, boundary_hash: "0123456789abcdef" },
    });
    const [, effects] = command(model, ports(), { kind: "control", link_id: "guest_1", wire });
    const observed = effects.find((effect) => effect.kind === "observe_hash");
    expect(observed?.kind === "observe_hash" ? observed.tick : undefined).toBe(30);
    expect(observed?.kind === "observe_hash" ? observed.hash : undefined).toBe("0123456789abcdef");
  });
});

describe("online match model full time and result", () => {
  function reachFullTime(model: OnlineMatchModel<FakeCoordinatorState, FixtureRequest>) {
    let next = model;
    [next] = command(next, ports(), { kind: "confirmed", steps: [step(0)] });
    return command(next, ports(), {
      kind: "full_time",
      final_tick: 90,
      home_score: 2,
      away_score: 1,
      final_hash: "0123456789abcdef",
    });
  }

  it("walks to full time and reports the simulation's own result", () => {
    const [finished, effects] = reachFullTime(harness().host);
    expect(publishedPhases(effects).join(",")).toBe("full_time");
    expect(sent(effects, "result_ack")).toBe(true);
    expect(finished.phase).toBe("full_time");
    expect(finished.score.home).toBe(2);
    expect(finished.score.away).toBe(1);
  });

  it("does not commit a result while any peer has not acknowledged it", () => {
    const [finished] = reachFullTime(harness().host);
    expect(finished.result).toBeUndefined();
    expect(finished.phase).toBe("full_time");
    expect(ended(finished)).toBe(false);
  });

  it("makes a guest report its own result rather than accept one on faith", () => {
    const [finished, effects] = reachFullTime(harness().guest);
    expect(publishedPhases(effects).length).toBe(0);
    expect(finished.phase).toBe("full_time");
    const localResult = finished.coordinator.local_result;
    expect(localResult?.home_score).toBe(2);
    expect(localResult?.away_score).toBe(1);
    expect(localResult?.final_hash).toBe("0123456789abcdef");
    expect(finished.result).toBeUndefined();
  });

  it("ignores a second full time", () => {
    const [first] = reachFullTime(harness().host);
    const [second, effects] = command(first, ports(), {
      kind: "full_time",
      final_tick: 91,
      home_score: 9,
      away_score: 9,
      final_hash: "ffffffffffffffff",
    });
    expect(effects.length).toBe(0);
    expect(second.score.home).toBe(2);
    expect(second.score.away).toBe(1);
  });
});

describe("online match model terminal reporting", () => {
  it("reports a netcode failure class into the session", () => {
    for (const testCase of [
      { failure: "late_input", reason: "late_input" },
      { failure: "desync", reason: "hash_mismatch" },
      { failure: "input_channel", reason: "input_channel_failure" },
    ]) {
      const model = harness().host;
      const [next] = command(model, ports(), {
        kind: "failure",
        failure: testCase.failure,
        detail: "a driver terminal",
      });
      expect(next.terminal?.reason).toBe(testCase.reason);
      expect(next.phase).toBe("ended");
      expect(next.result).toBeUndefined();
      expect(exitRoute(next)).toBe("terminal");
    }
  });

  it("reports a lost star as the guest's own lost host link", () => {
    const model = harness().guest;
    const [next, effects] = command(model, ports(), {
      kind: "failure",
      detail: "the star transport is no longer connected",
    });
    expect(next.terminal?.reason).toBe("transport_lost");
    expect(next.phase).toBe("ended");
    expect(publishedPhases(effects).length).toBe(0);
  });

  it("ends a host session explicitly when it cannot name the lost peer", () => {
    const model = harness().host;
    const [next, effects] = command(model, ports(), {
      kind: "failure",
      detail: "the star transport is no longer connected",
    });
    // The driver folds every link event into one status, so a host has no
    // peer id to disconnect. It ends the session rather than guessing.
    expect(next.terminal?.reason).toBe("local_abort");
    expect(sent(effects, "abort")).toBe(true);
    expect(next.phase).toBe("ended");
  });

  it("keeps the first terminal reason when late traffic follows it", () => {
    const model = harness().host;
    const [ended_] = command(model, ports(), { kind: "failure", failure: "late_input" });
    const [after, effects] = command(ended_, ports(), { kind: "confirmed", steps: [step(0)] });
    expect(effects.length).toBe(0);
    expect(after.terminal?.reason).toBe("late_input");
  });
});
